File: vapor-app/src/main/java/com/tevinjeffrey/vapor/okcloudapp/ProgressiveStringRequestBody.java
package com.tevinjeffrey.vapor.okcloudapp;

import com.squareup.okhttp.MediaType;
import com.squareup.okhttp.internal.Util;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.Charset;

import okio.BufferedSink;

public class ProgressiveStringRequestBody extends CloudAppRequestBody {

    public static final int DEFAULT_BUFFER_SIZE = 4096;

    private ProgressListener listener;
    private final InputStream inputStream;
    private final byte[] bytes;
    private final String name;

    public ProgressiveStringRequestBody(final String string, String name, ProgressListener listener) {
        Charset charset = Util.UTF_8;
        this.bytes = string.getBytes(charset);
        this.name = name == null? string.substring(0, string.length() < 18? string.length(): 18).trim() + ".txt" : name; //Sorry
        this.inputStream = new ByteArrayInputStream(bytes);
        this.listener = listener;
    }

    @Override
    public long contentLength() {
        return bytes.length;
    }

    @Override
    public MediaType contentType() {
        return MediaType.parse("text/plain");
    }

    @Override
    public void writeTo(BufferedSink sink) throws IOException {
        long fileLength = contentLength();
        byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];
        long total = 0;
        try {
            int read;
            while ((read = inputStream.read(buffer)) != -1) {
                this.listener.onProgress(total, fileLength);
                total += read;
                sink.write(buffer, 0, read);
            }
        } finally {
            inputStream.close();
        }
    }

    @Override
    public String getFileName() {
        return name;
    }
}

File: vapor-app/src/main/java/com/tevinjeffrey/vapor/okcloudapp/ProgressiveStringRequestBody.kt
package com.tevinjeffrey.vapor.okcloudapp

import com.squareup.okhttp.MediaType
import com.squareup.okhttp.internal.Util

import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.nio.charset.Charset

import okio.BufferedSink

class ProgressiveStringRequestBody(string: String, name: String?, private val listener: ProgressListener) : CloudAppRequestBody() {
    private val DEFAULT_BUFFER_SIZE = 4096

    override val fileName: String
        get() = name

    private val inputStream: InputStream
    private val bytes: ByteArray
    private val name: String

    init {
        val charset = Util.UTF_8
        this.bytes = string.toByteArray(charset)
        this.name = name ?: string.substring(0, if (string.length < 18) string.length else 18).trim { it <= ' ' } + ".txt" //Sorry
        this.inputStream = ByteArrayInputStream(bytes)
    }

    override fun contentLength(): Long {
        return bytes.size.toLong()
    }

    override fun contentType(): MediaType {
        return MediaType.parse("text/plain")
    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        val fileLength = contentLength()
        val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
        var total: Long = 0
        try {
            var read: Int = 0
            while (read != -1) {
                read = inputStream.read(buffer)
                this.listener.onProgress(total, fileLength)
                total += read.toLong()
                sink.write(buffer, 0, read)
            }
        } finally {
            inputStream.close()
        }
    }
}
